### ISSUE TITLE
Bonsai bug/ inaccuracy: Quantity Take-Off Perimeter

### ISSUE THREAD

**Issue by PahlawanJoey** [IfcOpenShell/IfcOpenShell#5795]
Example object GUID: km0003jkam16VSa07fDX00

Perimeter using 'Perform Quantity Take-off' with the 'IfcOpenShell' method has a inaccurate perimeter compared to the 'Blender' Method.

![002](https://github.com/user-attachments/assets/e92d1e72-ea5e-4ab0-a97a-731d463e5cc6)
![001](https://github.com/user-attachments/assets/48ec0d12-1aaf-45b1-ab1d-7c373268aa0d)
[https://file.io/5wnVLt9W1Seb](url)

**Comment by PahlawanJoey** [comment]
On the latest stable version with Blender 4.2.1.

**Comment by theoryshaw** [comment]
Should the perimeter be `20172mm`, if you take into considerations the voids?  Or is this asking too much of the calculation engines? 
![image](https://github.com/user-attachments/assets/7d272bca-9cc6-4578-8cb3-8da88950817b)

**Comment by sboddy** [comment]
From my read of https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/Qto_SpaceBaseQuantities.htm it looks like Blender is doing _NetPerimeter_ and IfcOpenShell is _maybe_ doing _GrossPerimeter_ and then they both call it simply Perimeter. The file is no longer available, but is that notch out the side a void, or with extra vertices in the slab geometry? Because NetPerimeter would ignore voids I think, making the Blender figure correct. GrossPerimeter (IfcOpenShell) looks to be not only counting the edge notch, but also all the inner void lengths too.

**Comment by sboddy** [comment]
From my read of https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/Qto_SlabBaseQuantities.htm 
> Perimeter measured along the outer boundaries of the slab. Only given, if the slab is prismatic (constant thickness).

It looks like Blender is doing a net Perimeter (ignore all voids) calculation and IfcOpenShell is doing a gross Perimeter (include voids) calculation and then they both call it Perimeter. The file is no longer available, but assuming that notch out the side is a void that is ignored for net calcs, this would make the Blender figure correct. While not explicit, the way the paragraph says "outer boundaries" would make me pick the 19,612 figure as the right one. It is "technically correct - the best kind of correct."

**Comment by PahlawanJoey** [comment]
> Should the perimeter be `20172mm`, if you take into considerations the voids?  Or is this asking too much of the calculation engines? 
> 
> ![image](https://github.com/user-attachments/assets/7d272bca-9cc6-4578-8cb3-8da88950817b)
> 
> 

Yes I agree, the perimeter should be including the void in your visualisation! How did you visualise that in Bonsai, or did you manually annotate it?

**Comment by PahlawanJoey** [comment]
> From my read of https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/Qto_SlabBaseQuantities.htm 
> 
> > Perimeter measured along the outer boundaries of the slab. Only given, if the slab is prismatic (constant thickness).
> 
> 
> 
> It looks like Blender is doing a net Perimeter (ignore all voids) calculation and IfcOpenShell is doing a gross Perimeter (include voids) calculation and then they both call it Perimeter. The file is no longer available, but assuming that notch out the side is a void that is ignored for net calcs, this would make the Blender figure correct. While not explicit, the way the paragraph says "outer boundaries" would make me pick the 19,612 figure as the right one. It is "technically correct - the best kind of correct."

Yes but it guess the values are both incorrect, the value in @theoryshaw his screenshot is what I would say is the correct perimeter. Here is a link to the model again, should be valid for 30 days otherwise just reply.

Also perhaps good to mention, there are no voids in the IFC model if I'm not mistaken. So the perimeter should consider the geometry instead?

https://easyupload.io/urqikw

**Comment by sboddy** [comment]
@PahlawanJoey The Plaatnummer 419 in what you provided didn't look like what @theoryshaw diagrammed. So I selected 420 which looked to have similar holes and a similar notch on the side. I toggled Voids (Alt-O) and lo and behold, every cut through the plate is a void. I arrowed _some_ of them here:
![image](https://github.com/user-attachments/assets/e938217e-f237-4f0a-9215-50ec987be4d8)
As to what is correct, if the spec says that voids are ignored for calculating the perimeter, then that does not mean "Oh, unless it's on an outside edge!". Because what if another software has followed the spec and now they disagree.

**Comment by sboddy** [comment]
Oh, and this might be a clue too ;-)
![image](https://github.com/user-attachments/assets/1b4f68b1-5cb7-4702-a8c6-9a3c9a12380d)

**Comment by theoryshaw** [comment]
just an fyi, you can change the suffix to `.txt` and upload file here...
[DYC_BD00_Breedplaat_Dakvloer.ifc.txt](https://github.com/user-attachments/files/17928732/DYC_BD00_Breedplaat_Dakvloer.ifc.txt)

**Comment by theoryshaw** [comment]
There's a glitch with `km0003jkam16VSa07fDX00` where the openings don't show when the file is loaded. 

video also shows how to use manual measure too. 

https://github.com/user-attachments/assets/b6571af5-4abd-4a6c-9edf-2d2cc5ac3dc4

**Comment by theoryshaw** [comment]
@PahlawanJoey maybe you can tell `IC-Prefab CAD` that their GlobalId's need fixing...
https://github.com/IfcOpenShell/IfcOpenShell/issues/5808#issuecomment-2503021494

**Comment by PahlawanJoey** [comment]
> km0003jkam16VSa07fDX00

Bug has been reported to IC-Prefab CAD.
